### ISSUE TITLE
Clang side expression warning

### ISSUE THREAD

**Issue by roigcarlo** [KratosMultiphysics/Kratos#399]
Mainly for you @RiccardoRossi 

The warning `-Wpotentially-evaluated-expression` present in `model_part_io.cpp`:

```
warning: expression with side effects will be evaluated despite being used as
      an operand to 'typeid' [-Wpotentially-evaluated-expression]
                if ((std::type_index(typeid(*itElemPrevious)) == std::type_index(typeid(*itElemCurrent))) && (previous_number_nodes =...
```

From here:

```C++
const std::type_info& type_component = typeid(*(i_comp->second));
if (std::type_index(typeid(*(rThisElements.begin()))) == std::type_index(type_component) &&
    (element_num_nodes == (i_comp->second)->GetGeometry().size()) &&
    (element_dimension == (i_comp->second)->GetGeometry().WorkingSpaceDimension())
)
```

Can be fixed using:

```C++
const auto i_comp_content = *(i_comp->second);
const auto rThisElements_content = *(rThisElements.begin());

const auto & i_comp_component = typeid(i_comp_content);
const auto & rThisElements_component = typeid(rThisElements_content);

if (std::type_index(rThisElements_component) == std::type_index(i_comp_component) &&
    (element_num_nodes == (i_comp->second)->GetGeometry().size()) &&
    (element_dimension == (i_comp->second)->GetGeometry().WorkingSpaceDimension())
)
```

Should we go ahead and fix this in the clang warnings branch?

**Comment by RiccardoRossi** [comment]
Hi Charlie,

    If I understand you are copy constructing all the components right? If it is so, then I would say that's very expensive fix for a warning. 

In this link they propose to do the comparison in a helper function

http://trac.wxwidgets.org/ticket/16968

What do you think?

**Comment by roigcarlo** [comment]
Fine, as well as long as we don't access the pointer inside the `typeid` t should do the trick 👍

**Comment by RiccardoRossi** [comment]
Well the point is not to make a copy...

Riccardo

**Comment by roigcarlo** [comment]
Ha, I see, yep. Missed that  :flushed:. Then with the function we can avoid the copy, but as they were only pointers i didn't give itmuch importance. 

Thx por pointing it out

**Comment by pooyan-dadvand** [comment]
It will be addressed with the solution of #188. So I will close it.
